platforms/org.eclipse.mita.platform.xdk110/src/org/eclipse/mita/platform/xdk110/io/SDCardGenerator.xtend

```diff
@@ -0,0 +1,233 @@
+		val result = StaticValueInferrer.infer(ModelUtils.getArgumentValue(instance, instance.sizeName), []);
+		if(result instanceof Integer) {
+			return codeFragmentProvider.create('''«result»''');
+		}
+		else {
+			return codeFragmentProvider.create('''-1''');	
+		}
+		
+	}
+	
+	static def String getSizeName(SignalInstance instance) {
+		if(instance.instanceOf.name.startsWith("appendingFile")) {
+			return "blockSize";
+		} 
+		else {
+			return "fileSize";
+		}
+	}
+	
+	override generateEnable() {
+		codeFragmentProvider.create('''
+			Retcode_T retcode = RETCODE_OK;
+			retcode = SDCardDriver_Initialize();
+			if (RETCODE_OK == retcode)
+			{
+				if (SDCARD_INSERTED != SDCardDriver_GetDetectStatus())
+				{
+					«loggingGenerator.generateLogStatement(LogLevel.Error, "SD card was not detected for Storage")»
+					retcode = RETCODE(RETCODE_SEVERITY_WARNING, RETCODE_STORAGE_SDCARD_NOT_AVAILABLE);
+				}
+			}
+			if (RETCODE_OK == retcode)
+			{
+				retcode = SDCardDriver_DiskInitialize(STORAGE_SDCARD_DRIVE_NUMBER); /* Initialize SD card */
+			}
+			if (RETCODE_OK == retcode)
+			{
+				/* Initialize file system */
+				if (FR_OK != f_mount(&StorageSDCardFatFSObject, STORAGE_DEFAULT_LOGICAL_DRIVE, STORAGE_SDCARD_FORCE_MOUNT))
+				{
+					retcode = RETCODE(RETCODE_SEVERITY_ERROR, RETCODE_STORAGE_SDCARD_MOUNT_FAILED);
+				}
+			}
+			return retcode;
+		''')
+	}
+	
+	override generateSignalInstanceGetter(SignalInstance sigInst, String valueVariableName) {
+		val data = sigInst.dataAccessor(valueVariableName);
+		val len = sigInst.getSize;
+		val filename = sigInst.filenameAccessor(valueVariableName);
+		val fileSeekIndex = if(sigInst.instanceOf.name.startsWith("persistentFile")) {
+			    codeFragmentProvider.create('''«sigInst.name»FilePosition''');
+			} 
+			else {
+			    codeFragmentProvider.create('''0''');
+			}
+		codeFragmentProvider.create('''
+			Retcode_T retcode = RETCODE_OK;
+			FRESULT sdCardReturn = FR_OK, fileOpenReturn = FR_OK;
+			FILINFO sdCardFileInfo;
+			#if _USE_LFN
+			sdCardFileInfo.lfname = NULL;
+			#endif
+			FIL fileReadHandle;
+			UINT bytesRead;
+			
+			sdCardReturn = f_stat(«filename», &sdCardFileInfo);
+			if (FR_OK == sdCardReturn)
+			{
+				fileOpenReturn = f_open(&fileReadHandle, «filename», FA_OPEN_EXISTING | FA_READ);
+			}
+			if ((FR_OK == sdCardReturn) && (FR_OK == fileOpenReturn))
+			{
+			    sdCardReturn = f_lseek(&fileReadHandle, «fileSeekIndex»);
+			}
+			if(«fileSeekIndex» >= sdCardFileInfo.fsize)
+			{
+				return EXCEPTION_ENDOFFILEEXCEPTION;
+			}
+			if ((FR_OK == sdCardReturn) && (FR_OK == fileOpenReturn))
+			{
+			    sdCardReturn = f_read(&fileReadHandle, «data», «len», &bytesRead); /* Read a chunk of source file */
+			}
+			if ((FR_OK == sdCardReturn) && (FR_OK == fileOpenReturn))
+			{
+				«IF sigInst.instanceOf.name.startsWith("appendingFile")»
+				«sigInst.name»FilePosition += bytesRead;
+				«ENDIF»
+			    sdCardReturn = f_close(&fileReadHandle);

```

Comment:
Inconsistent indentation here and `f_close` should probably always happen if `fileOpenReturn == FR_OK`, since we want to keep the file closed in between method calls.